mirrord/operator/src/client.rs

```diff
@@ -1,557 +1,709 @@
-        self.client_certificate
-            .as_ref()
-            .map(|cert| {
-                cert.encode_der()
-                    .map(|bytes| general_purpose::STANDARD.encode(bytes))
-            })
-            .transpose()
+    /// Tries to fetch [`MirrordOperatorCrd`] from the cluster and create a new instance of this
+    /// API.
+    ///
+    /// If fetching the resource fails, an extra discovery step is made to determine whether the
+    /// operator is installed. If this extra step proves that the operator is installed, an
+    /// error is returned. Otherwise, [`None`] is returned.
+    #[tracing::instrument(level = Level::TRACE, skip_all, err)]
+    pub async fn try_new<R>(
+        config: &LayerConfig,
+        reporter: &mut R,
+    ) -> OperatorApiResult<Option<Self>>
+    where
+        R: Reporter,
+    {
+        let base_config = Self::base_client_config(config).await?;
+        let client = Client::try_from(base_config.clone())
+            .map_err(KubeApiError::from)
+            .map_err(OperatorApiError::CreateKubeClient)?;
+
+        let operator: Result<MirrordOperatorCrd, _> =
+            Api::all(client.clone()).get(OPERATOR_STATUS_NAME).await;
+
+        let error = match operator {
+            Ok(operator) => {
+                reporter.set_operator_properties(AnalyticsOperatorProperties {
+                    client_hash: None,
+                    license_hash: operator
+                        .spec
+                        .license
+                        .fingerprint
+                        .as_deref()
+                        .map(AnalyticsHash::from_base64),
+                });
+
+                return Ok(Some(Self {
+                    client,
+                    client_cert: None,
+                    base_config,
+                    operator,
+                }));
+            }
+
+            Err(error @ kube::Error::Api(..)) => {
+                match discovery::operator_installed(&client).await {
+                    Ok(false) | Err(..) => {
+                        return Ok(None);
+                    }
+                    Ok(true) => error,
+                }
+            }
+
+            Err(error) => error,
+        };
+
+        Err(OperatorApiError::KubeError {
+            error,
+            operation: OperatorOperation::FindingOperator,
+        })
     }
 
-    fn set_operator_properties<R: Reporter>(&self, analytics: &mut R) {
-        let client_hash = self
-            .client_certificate
-            .as_ref()
-            .map(|cert| cert.public_key_data())
-            .as_deref()
-            .map(AnalyticsHash::from_bytes);
+    /// Prepares client [`Certificate`] to be sent in all subsequent requests to the operator.
+    /// In case of failure, state of this API instance does not change.
+    #[tracing::instrument(level = Level::TRACE, skip(reporter), err)]
+    pub async fn prepare_client_cert<R>(&mut self, reporter: &mut R) -> OperatorApiResult<()>
```

Comment:
Having some sort of middleware layer for `kube::Client` using their builder can be an elegant solution for keeping the certificate up to date for each request without having to call it after creating the operator client though there might be some complications and double clienting, alternatively there is maybe some merit in making it a type state machine where you have `OperatorApi<T = Unauthorised>` and only post upgrade of `prepare_client_cert` you have `OperatorApi<WithCertificate>` that has all the methods that require credentials

Reply:
Type state machine sound cool. I don't want to play with adding middleware though, Aviram added extra headers field to `kube::Config` and imho it's simpler

Reply:
Updated ^^